src/operator/numpy/linalg/np_gesvd-inl.h

```diff
@@ -0,0 +1,298 @@
+      denom = L[offl + i];
+      if (denom < eps) denom = eps;
+      for (int j = 0; j < n; ++j) {
+        dA[offda + i * ldda + j] /= denom;
+      }
+    }
+  }
+};
+
+struct GesvdBackHelper_G1 {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int k, int m, int n, DType* X, int ldx,
+                                  DType* L, int ldl) {
+    const int offl(k * ldl);
+    const int offx(k * m * ldx);
+    DType numer(0.0);
+    for (int i = 0; i < m; ++i) {
+      for (int j = 0; j < m; ++j) {
+        numer = L[offl + j];
+        X[offx + i * ldx + j] *= numer;
+      }
+    }
+  }
+};
+
+struct GesvdBackHelper_G2 {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int k, int m, int n, DType* X, int ldx,
+                                  DType* L, int ldl, DType* dL, int lddl,
+                                  DType* dA, int ldda, DType* V, int ldv) {
+    const int offx(k * m * ldx);
+    const int offl(k * ldl);
+    const int offdl(k * lddl);
+    const int offda(k * m * ldda);
+    const int offv(k * m * ldv);
+    const DType eps(gesvd_back_helper_eps(X));
+    DType denom1(0.0), denom2(0.0), elem(0.0);
+
+    for (int i = 0; i < m - 1; ++i) {
+      for (int j = i + 1; j < m; ++j) {
+        denom1 = L[offl + i] - L[offl + j];
+        denom2 = L[offl + i] + L[offl + j];
+        if (denom1 < eps) denom1 = eps;
+        if (denom2 < eps) denom2 = eps;
+        elem = (X[offx + i * ldx + j] - X[offx + j * ldx + i]) / denom1 / denom2;
+        X[offx + i * ldx + j] = elem * L[offl + j];
+        X[offx + j * ldx + i] = elem * L[offl + i];
+      }
+    }
+    for (int i = 0; i < m; ++i) {
+      elem = DType(0.0);
+      for (int j = 0; j < n; ++j) {
+        elem += dA[offda + i * ldda + j] * V[offv + i * ldv + j];
+      }
+      elem = -elem + dL[offdl + i];
+      X[offx + i * ldx + i] = elem;
+    }
+  }
+};
+
+struct gesvd_backward {
+  template<typename xpu, typename DType>
+  static void op(const Tensor<xpu, 3, DType>& dUT,
+                 const Tensor<xpu, 2, DType>& dL,
+                 const Tensor<xpu, 3, DType>& dV,
+                 const Tensor<xpu, 3, DType>& UT,
+                 const Tensor<xpu, 2, DType>& L,
+                 const Tensor<xpu, 3, DType>& V,
+                 const Tensor<xpu, 3, DType>& dA,
+                 const Tensor<xpu, 3, DType>& tempMs,
+                 const Tensor<xpu, 3, DType>& tempMr,
+                 Stream<xpu>* s, const nnvm::NodeAttrs& attrs) {
+    // Backward of (UT, L, V) = gesvd(A)
+    using namespace mxnet_op;
+    if (dA.dptr_ != dV.dptr_) {
+      Copy(dA, dV, s);
+    }
+    // From here on, we work on dA only
+
+    // Need temporal space, same shape as dUT
+    // invdV:
+    Kernel<GesvdBackHelper_dV, xpu>::Launch
+      (s, dA.size(0), dA.size(1), dA.size(2), L.dptr_, L.stride_, dA.dptr_, dA.stride_);
+
+    // G1:
+    // This copy is just to make sure there are no invalid values (NaN, infinity) in tempM
+    Copy(tempMs, dUT, s);
```

Comment:
May be cleaner to fill with zero instead of just copying some arbitrary data. I think there is a specific method in mshadow or elsewhere to fill a tensor with such values (if not, copying like here is likely the best way)

Reply:
The only way I know to fill tensor with zero is to:
```
    tempMs.FlatTo1D() = 0;
    tempMr.FlatTo1D() = 0;
```
But our old code (syevd) sticks to copying. Which do you think is better?

Reply:
The copy in syevd is [here](https://github.com/apache/incubator-mxnet/blob/67daae7eff1ef06254693bc1bc30bbfc23b2cf99/src/operator/tensor/la_op-inl.h#L861)

Reply:
Aha, so it seems I am to blame for that. But it is fine

Reply:
I changed the copying to filling with zeros.